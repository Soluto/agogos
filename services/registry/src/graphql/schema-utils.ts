import {
    DefinitionNode,
    DocumentNode,
    FieldDefinitionNode,
    InputValueDefinitionNode
} from "graphql";
import {
    clone,
    equals,
    flatten,
    groupBy,
    indexBy,
    map,
    pipe,
    unionWith,
    uniq,
    values
} from "ramda";

function validateSameFieldDefs(
    fieldDefA: FieldDefinitionNode | InputValueDefinitionNode,
    fieldDefB: FieldDefinitionNode | InputValueDefinitionNode
) {
    const { directives: _, ...fieldDefWithNoDirectivesA } = fieldDefA;
    const { directives: __, ...fieldDefWithNoDirectivesB } = fieldDefB;
    if (!equals(fieldDefWithNoDirectivesA, fieldDefWithNoDirectivesB)) {
        throw new Error(`Same field definition should be identical except for directives - field ${
            fieldDefA.name.value
            }`);
    }
}

function mergeSameFieldDefinitions(
    fieldDefinitions: ReadonlyArray<
        FieldDefinitionNode | InputValueDefinitionNode
    >
): ReadonlyArray<FieldDefinitionNode | InputValueDefinitionNode> {
    if (fieldDefinitions.length === 0) { return []; }
    if (fieldDefinitions.length === 1) { return fieldDefinitions; }

    const [fieldDefA, fieldDefB, ...rest] = fieldDefinitions;
    validateSameFieldDefs(fieldDefA, fieldDefB);

    const { directives: _, ...otherFields } = fieldDefA;
    const fieldWithMergedDirectives = {
        ...otherFields,
        directives: uniq([
            ...(fieldDefA.directives || []),
            ...(fieldDefB.directives || [])
        ])
    };

    return mergeSameFieldDefinitions([fieldWithMergedDirectives, ...rest]);
}

const mergeFields = (
    l: ReadonlyArray<FieldDefinitionNode | InputValueDefinitionNode>,
    r: ReadonlyArray<FieldDefinitionNode | InputValueDefinitionNode>
) =>
    pipe(
        groupBy(
            (fieldDef: FieldDefinitionNode | InputValueDefinitionNode) =>
                fieldDef.name.value
        ),
        values,
        map(mergeSameFieldDefinitions),
        flatten
    )([...l, ...r]);

export function mergeDocuments(schemas: DocumentNode[]) {
    return clone(schemas).reduce((acc: DocumentNode, next: DocumentNode) => {
        const index = indexBy((x: DefinitionNode) => {
            if (
                x.kind === "ObjectTypeDefinition" ||
                x.kind === "UnionTypeDefinition" ||
                x.kind === "InputObjectTypeDefinition"
            ) {
                return `${x.kind}_${x.name.value}`;
            }
            return "";
        })(acc.definitions);

        const get = (kind: string, value: string) => index[`${kind}_${value}`];

        next.definitions.forEach(x => {
            if (
                (x.kind !== "ObjectTypeDefinition" &&
                    x.kind !== "UnionTypeDefinition" &&
                    x.kind !== "InputObjectTypeDefinition") ||
                !get(x.kind, x.name.value)
            ) {
                (acc.definitions as any).push(x);
                return;
            }

            const ref = get(x.kind, x.name.value);
            if (
                x.kind === "ObjectTypeDefinition" ||
                x.kind === "InputObjectTypeDefinition"
            ) {
                (ref as any).directives = uniq([
                    ...(ref as any).directives,
                    ...(x.directives || [])
                ]);
                (ref as any).fields = mergeFields((ref as any).fields, x.fields || []);

                if (x.kind === "ObjectTypeDefinition") {
                    (ref as any).interfaces = uniq([
                        ...(ref as any).interfaces,
                        ...(x.interfaces || [])
                    ]);
                }
                return;
            }

            if (x.kind === "UnionTypeDefinition") {
                (ref as any).types = unionWith(equals, (ref as any).types, x.types || []);
                return;
            }
        });

        return acc;
    });
}
