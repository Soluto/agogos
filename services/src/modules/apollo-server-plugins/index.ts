import { PluginDefinition } from 'apollo-server-core';
import { transformApolloServerPlugins as applyPluginsForApolloServerPlugins } from '../plugins';
import { createBasicPolicyPlugin } from './base-policy';
import { createLoggingPlugin } from './logging';
import { createMetricsPlugin } from './metrics';

export default function getPlugins(): PluginDefinition[] {
  const baseApolloServerPlugins = [createBasicPolicyPlugin, createLoggingPlugin, createMetricsPlugin];
  const plugins = applyPluginsForApolloServerPlugins(baseApolloServerPlugins);
  return plugins;
}
