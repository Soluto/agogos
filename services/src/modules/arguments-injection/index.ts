import getJwt from './request-context-jwt';
import evaluate from './arguments-evaluation';
import { inject, deepInject } from './arguments-injection';

export { getJwt, inject, deepInject, evaluate as injectArgs };
