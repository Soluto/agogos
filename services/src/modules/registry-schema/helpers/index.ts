import proceedPolicy from './policy-helper';

export { proceedPolicy };
