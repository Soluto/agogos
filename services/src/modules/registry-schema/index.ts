import typeDefs from './typedefs';
import resolvers from './resolvers';
export * from './types';

export { typeDefs, resolvers };
