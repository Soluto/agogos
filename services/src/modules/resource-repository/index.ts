export * from './types';

export { ResourceRepository } from './resource-repository';
export { CompositeResourceRepository } from './composite';
export { S3ResourceRepository } from './s3-repository';
export { FileSystemResourceRepository } from './fs-repository';
export { pollForUpdates } from './stream';
export { applyResourceGroupUpdates, applyResourceGroupDeletions } from './actions';
export { default as getResourceRepository } from './get-resource-repository';
