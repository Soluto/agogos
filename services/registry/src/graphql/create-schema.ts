import gql from "graphql-tag";
import { mergeDocuments } from "./schema-utils";
import { DocumentNode } from "graphql";

type GqlSource = {
  source: string;
  gql: string;
};

export type GqlSources = { [source: string]: string };

function parseGqlSource(gqlSource: GqlSource) {
  try {
    return gql(gqlSource.gql);
  } catch (e) {
    throw {
      message: `Failed to load schema for source -  ${gqlSource.source} - ${
        e.message
      }`,
      schema: gqlSource,
      error: e
    };
  }
}

function mergeAllDocuments(docs: DocumentNode[]) {
  try {
    return mergeDocuments(docs);
  } catch (e) {
    throw {
      message: `Failed to merge schemas - ${e.message}`,
      schemas: docs,
      error: e
    };
  }
}

export const makeGqlDocumentFromGqlSources = (gqlSources: GqlSources) => {
  const documentNodes = Object.entries(gqlSources).map(([source, gql]) =>
    parseGqlSource({ source, gql })
  );
  return mergeAllDocuments(documentNodes);
};
