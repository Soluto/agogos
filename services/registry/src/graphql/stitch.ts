import {DocumentNode, parse} from 'graphql';
import {mergeDocuments} from './schema-utils';
import {AggObjsByName} from '../sync/sync-service';
import {SchemaConfig} from '../sync/object-types';

function parseGqlSource(schemaName: string, schema: string) {
    try {
        return parse(schema);
    } catch (e) {
        throw {
            error: e,
            message: `Failed to load schema for source -  ${schemaName} - ${e.message}`,
            schema: {schemaName, schema},
        };
    }
}

function mergeAllDocuments(docs: DocumentNode[]) {
    try {
        return mergeDocuments(docs);
    } catch (e) {
        throw {
            error: e,
            message: `Failed to merge schemas - ${e.message}`,
            schemas: docs,
        };
    }
}

export const stitchSchemas = (gqlSchemas: AggObjsByName) => {
    const documentNodes = Object.entries(gqlSchemas).map(([schemaName, schema]: [string, SchemaConfig]) =>
        parseGqlSource(schemaName, schema.definition)
    );
    return mergeAllDocuments(documentNodes);
};
