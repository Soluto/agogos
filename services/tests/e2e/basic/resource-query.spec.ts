import { print } from 'graphql';
import { gql } from 'apollo-server-core';
import { GraphQLClient } from 'graphql-request';
import GraphQLErrorSerializer from '../../utils/graphql-error-serializer';
import { getToken } from '../../helpers/get-token';
import { ResourceMetadataInput } from '../../../src/modules/registry-schema';
import { PolicyDefinition, PolicyType } from '../../../src/modules/resource-repository';
import { RegistryMutationResponse, updatePoliciesMutation } from '../../helpers/registry-request-builder';
import { sleep } from '../../helpers/utility';

const gatewayClient = new GraphQLClient('http://localhost:8080/graphql');
const registryClient = new GraphQLClient('http://localhost:8090/graphql');

describe('Basic flow', () => {
  const query = print(gql`
    query($metadata: ResourceMetadataInput!, $fromGatewayResources: Boolean) {
      policy(metadata: $metadata, fromGatewayResources: $fromGatewayResources) {
        code
      }
    }
  `);

  const metadata: ResourceMetadataInput = {
    namespace: 'resource-query',
    name: 'policy1',
  };

  beforeAll(async () => {
    expect.addSnapshotSerializer(GraphQLErrorSerializer);

    const accessToken = await getToken();
    gatewayClient.setHeader('Authorization', `Bearer ${accessToken}`);
  });

  test('Query before adding policy', async () => {
    const registryResult = await registryClient.request<{ policy?: PolicyDefinition }>(query, { metadata });
    expect(registryResult.policy).toBeNull();

    const gatewayResult = await registryClient.request<{ policy?: PolicyDefinition }>(query, {
      metadata,
      fromGatewayResources: true,
    });
    expect(gatewayResult.policy).toBeNull();
  });

  test('Add policy and query', async () => {
    const policies: PolicyDefinition[] = [
      {
        metadata,
        type: PolicyType.opa,
        code: `default allow = false`,
      },
    ];

    const updateResponse = await registryClient.request<RegistryMutationResponse>(updatePoliciesMutation, {
      policies,
    });
    expect(updateResponse.result.success).toBeTruthy();

    // Wait for gateway to update before next tests
    await sleep(Number(process.env.WAIT_FOR_REFRESH_ON_GATEWAY) | 1500);

    const registryResult = await registryClient.request<{ policy?: PolicyDefinition }>(query, { metadata });
    expect(registryResult.policy).toMatchSnapshot();

    const gatewayResult = await registryClient.request<{ policy?: PolicyDefinition }>(query, {
      metadata,
      fromGatewayResources: true,
    });
    expect(gatewayResult.policy).toMatchSnapshot();
  }, 10000);

  test('Get gateway resource added by plugin', async () => {
    const gatewayOnlyResourceMetadata: ResourceMetadataInput = {
      namespace: 'plugins',
      name: 'whole-rg-policy',
    };

    const registryResult = await registryClient.request<{ policy?: PolicyDefinition }>(query, {
      metadata: gatewayOnlyResourceMetadata,
    });
    expect(registryResult.policy).toBeNull();

    const gatewayResult = await registryClient.request<{ policy?: PolicyDefinition }>(query, {
      metadata: gatewayOnlyResourceMetadata,
      fromGatewayResources: true,
    });
    expect(gatewayResult.policy).toMatchSnapshot();
  });
});
