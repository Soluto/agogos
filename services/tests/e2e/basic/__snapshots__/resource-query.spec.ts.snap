// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Basic flow Add policy and query 1`] = `
Object {
  "code": "default allow = false",
}
`;

exports[`Basic flow Add policy and query 2`] = `
Object {
  "code": "default allow = false",
}
`;

exports[`Basic flow Get gateway resource added by plugin 1`] = `
Object {
  "code": "
    default allow = true
  ",
}
`;
